File: agent-service/app/agent/document_strategy.py
from __future__ import annotations

import json
from typing import Any

from app.schemas.document_contracts import (
    DOCUMENT_ARTIFACTS,
    DOCUMENT_EVIDENCE_SOURCES,
    AiDocumentPlan,
    AiDocumentPlanSection,
    AiDocumentStrategy,
    DocumentArtifact,
    DocumentEvidenceSource,
)


DEFAULT_EDITOR_HINTS = [
    "Use Markdown tables for comparisons, parameters, matrices, and checklists.",
    "Use fenced code blocks with language tags for code, commands, and configuration.",
    "Use Mermaid code blocks for architecture, process, and sequence diagrams when a diagram clarifies the topic.",
    "Use Docmost callouts with :::info, :::warning, :::danger, or :::success for key notes and risks.",
    "Use :::details blocks for optional supplementary detail, not for mandatory content.",
    "Use Markdown images ![alt](url) when source or generated visuals materially help the reader.",
]

EVIDENCE_SOURCE_ALIASES: dict[str, DocumentEvidenceSource] = {
    "search": "web_search",
    "crawl": "web_crawl",
    "knowledge_base": "knowledge_search",
    "image": "generated_image",
    "image_generation": "generated_image",
}


def _normalize_artifacts(raw_values: Any) -> list[DocumentArtifact]:
    if not isinstance(raw_values, list):
        return []

    normalized: list[DocumentArtifact] = []
    for value in raw_values:
        candidate = str(value).strip()
        if candidate in DOCUMENT_ARTIFACTS and candidate not in normalized:
            normalized.append(candidate)  # type: ignore[arg-type]
    return normalized


def _normalize_evidence_sources(raw_values: Any) -> list[DocumentEvidenceSource]:
    if not isinstance(raw_values, list):
        return []

    normalized: list[DocumentEvidenceSource] = []
    for value in raw_values:
        candidate = EVIDENCE_SOURCE_ALIASES.get(str(value).strip(), str(value).strip())
        if (
            candidate in DOCUMENT_EVIDENCE_SOURCES
            and candidate not in normalized
        ):
            normalized.append(candidate)  # type: ignore[arg-type]
    return normalized


def empty_document_plan(strategy: AiDocumentStrategy | None = None) -> AiDocumentPlan:
    return normalize_document_plan({}, strategy)


def format_document_strategy(strategy: AiDocumentStrategy | dict[str, Any] | None) -> str:
    strategy = strategy or {}
    lines: list[str] = []

    if strategy.get("docType"):
        lines.append(f"Document type: {strategy['docType']}")
    if strategy.get("audience"):
        lines.append(f"Audience: {strategy['audience']}")

    for label, key in (
        ("Objectives", "objectives"),
        ("Required sections", "requiredSections"),
        ("Required artifacts", "requiredArtifacts"),
        ("Optional artifacts", "optionalArtifacts"),
        ("Review checks", "reviewChecks"),
    ):
        values = strategy.get(key) or []
        if not values:
            continue
        lines.append(f"{label}:")
        for value in values:
            lines.append(f"- {value}")

    editor_hints = strategy.get("editorSyntaxHints") or DEFAULT_EDITOR_HINTS
    lines.append("Editor syntax hints:")
    for hint in editor_hints:
        lines.append(f"- {hint}")

    return "\n".join(lines).strip()


def derive_visual_requirements(state: dict[str, Any]) -> list[DocumentArtifact]:
    text = " ".join(
        filter(
            None,
            [
                state.get("user_message", ""),
                json.dumps(state.get("document_strategy") or {}, ensure_ascii=False),
            ],
        )
    ).lower()

    hints: list[str] = []
    if any(token in text for token in ["流程", "架构", "workflow", "sequence", "architecture", "流程图", "时序"]):
        hints.append("mermaid")
    if any(token in text for token in ["对比", "compare", "矩阵", "参数", "checklist", "表格", "表"]):
        hints.append("table")
    if any(token in text for token in ["截图", "image", "图片", "screen", "界面", "示意图"]):
        hints.append("image")
    if any(token in text for token in ["代码", "script", "命令", "sql", "api", "curl", "yaml", "json"]):
        hints.append("code_block")
    return [artifact for artifact in DOCUMENT_ARTIFACTS if artifact in hints]


def normalize_document_plan(
    raw: Any,
    strategy: AiDocumentStrategy | dict[str, Any] | None = None,
) -> AiDocumentPlan:
    strategy = strategy or {}
    required_sections = strategy.get("requiredSections") or []
    required_artifacts = _normalize_artifacts(strategy.get("requiredArtifacts") or [])

    if not isinstance(raw, dict):
        raw = {}

    sections = raw.get("sections")
    if not isinstance(sections, list) or not sections:
        sections = [
            {
                "title": title,
                "goal": title,
                "artifacts": list(required_artifacts),
                "must_cover": [],
                "evidence": [],
            }
            for title in required_sections
        ]

    normalized_sections: list[AiDocumentPlanSection] = []
    for index, section in enumerate(sections, start=1):
        if not isinstance(section, dict):
            continue
        normalized_sections.append(
            {
                "id": section.get("id") or f"section-{index}",
                "title": str(section.get("title") or f"Section {index}"),
                "goal": str(section.get("goal") or section.get("title") or f"Section {index}"),
                "artifacts": _normalize_artifacts(section.get("artifacts") or []),
                "must_cover": list(dict.fromkeys(section.get("must_cover") or [])),
                "evidence": _normalize_evidence_sources(section.get("evidence") or []),
            }
        )

    return {
        "doc_type": str(raw.get("doc_type") or strategy.get("docType") or "general-document"),
        "audience": str(raw.get("audience") or strategy.get("audience") or ""),
        "required_artifacts": _normalize_artifacts(
            raw.get("required_artifacts") or required_artifacts
        ),
        "sections": normalized_sections,
    }

File: apps/server/src/ee/ai/document-strategy.ts
import type { AiDocumentArtifact } from './document-plan';

export type AiIntentRoute =
  | 'selection_edit'
  | 'document_transform'
  | 'document_create';
export type AiIntentScope =
  | 'selection'
  | 'uploaded_document'
  | 'current_page'
  | 'blank_page';
export type AiSourcePolicy =
  | 'preserve_source'
  | 'transform_source'
  | 'create_new';
export type AiLengthPolicy = 'preserve' | 'compress' | 'expand';

export interface AiDocumentStrategy {
  templateKey: string;
  docType: string;
  audience: string;
  objectives: string[];
  requiredArtifacts: AiDocumentArtifact[];
  optionalArtifacts: AiDocumentArtifact[];
  requiredSections: string[];
  reviewChecks: string[];
  editorSyntaxHints: string[];
  intentRoute: AiIntentRoute;
  scope: AiIntentScope;
  sourcePolicy: AiSourcePolicy;
  lengthPolicy: AiLengthPolicy;
  prioritizeUserInstructions: boolean;
}

const COMMON_EDITOR_HINTS = [
  'Use Markdown tables when comparing options, parameters, or checklists.',
  'Use fenced code blocks with language tags for runnable code or commands.',
  'Use Mermaid blocks for architecture, process, or sequence explanations when a diagram materially improves understanding.',
  'Use Docmost callouts with :::info / :::warning / :::danger / :::success for tips, risks, and key takeaways.',
  'Use :::details blocks for collapsible supplementary material, not for core steps or conclusions.',
  'Use images with standard Markdown syntax ![alt](url) and explain why the image matters in surrounding text.',
];

const DEFAULT_STRATEGY: AiDocumentStrategy = {
  templateKey: 'general',
  docType: 'general-document',
  audience: 'knowledge worker reading and editing in Docmost',
  objectives: [
    'Answer the user request directly instead of padding with generic prose.',
    'Prefer evidence-backed statements and clearly mark uncertain claims.',
  ],
  requiredArtifacts: [],
  optionalArtifacts: ['table', 'code_block', 'callout'],
  requiredSections: [],
  reviewChecks: [
    'No filler introduction or conclusion.',
    'Each major section must have a clear purpose.',
    'Prefer compact structure over repetitive prose.',
  ],
  editorSyntaxHints: COMMON_EDITOR_HINTS,
  intentRoute: 'document_create',
  scope: 'blank_page',
  sourcePolicy: 'create_new',
  lengthPolicy: 'preserve',
  prioritizeUserInstructions: true,
};

const TEMPLATE_STRATEGIES: Record<string, Partial<AiDocumentStrategy>> = {
  'technical-doc': {
    docType: 'technical-documentation',
    audience: 'engineers and technical operators',
    objectives: [
      'Explain what the system does, how it works, and how to use it.',
      'Ground explanations in the provided materials and concrete examples.',
    ],
    requiredArtifacts: ['code_block'],
    optionalArtifacts: ['mermaid', 'table', 'callout', 'details'],
    requiredSections: ['overview', 'key concepts', 'usage or workflow'],
    reviewChecks: [
      'Technical claims should be specific and implementation-aware.',
      'Use at least one non-prose artifact when complexity warrants it.',
      'Examples should be runnable or concretely interpretable.',
    ],
  },
  'operation-manual': {
    docType: 'operations-manual',
    audience: 'frontline operators or non-technical users',
    objectives: [
      'Produce step-by-step instructions that can be executed without inference.',
      'Preserve or request visuals when they materially support a step.',
    ],
    requiredArtifacts: ['callout'],
    optionalArtifacts: ['table', 'image', 'details'],
    requiredSections: ['preparation', 'steps', 'verification', 'troubleshooting'],
    reviewChecks: [
      'Every critical step must include expected outcome or verification.',
      'Risky actions must include warnings.',
      'Use images when the source material contains relevant screenshots or diagrams.',
    ],
  },
  'meeting-notes': {
    docType: 'meeting-notes',
    audience: 'participants and stakeholders',
    objectives: [
      'Capture decisions, owners, and next steps clearly.',
      'Keep narrative concise and action-oriented.',
    ],
    requiredArtifacts: ['table'],
    optionalArtifacts: ['callout'],
    requiredSections: ['summary', 'discussion points', 'decisions', 'action items'],
    reviewChecks: [
      'Action items should include owners or placeholders for owners.',
      'Decisions and open questions must be distinguishable.',
    ],
  },
  requirements: {
    docType: 'requirements-analysis',
    audience: 'product, design, and engineering stakeholders',
    objectives: [
      'Translate user intent into concrete requirements and acceptance criteria.',
      'Make scope and non-goals explicit.',
    ],
    requiredArtifacts: ['table'],
    optionalArtifacts: ['mermaid', 'callout'],
    requiredSections: ['background', 'user scenarios', 'functional requirements', 'acceptance criteria'],
    reviewChecks: [
      'Requirements must be testable and not purely aspirational.',
      'Acceptance criteria should be explicit.',
    ],
  },
  report: {
    docType: 'research-report',
    audience: 'decision makers or analysts',
    objectives: [
      'Synthesize evidence into conclusions, not just excerpts.',
      'Show comparisons, trends, and implications clearly.',
    ],
    requiredArtifacts: ['table'],
    optionalArtifacts: ['mermaid', 'callout', 'image'],
    requiredSections: ['executive summary', 'findings', 'analysis', 'recommendations'],
    reviewChecks: [
      'Major claims should map to evidence or citations.',
      'Comparative statements should be normalized into tables when possible.',
    ],
  },
  prd: {
    docType: 'product-requirements-document',
    audience: 'product, design, engineering, and QA',
    objectives: [
      'Separate why, what, and how clearly.',
      'Produce implementation-ready requirements and validation criteria.',
    ],
    requiredArtifacts: ['table'],
    optionalArtifacts: ['mermaid', 'callout', 'details'],
    requiredSections: ['goals', 'user stories', 'functional design', 'acceptance criteria'],
    reviewChecks: [
      'Requirements must be testable.',
      'Risks, open questions, and dependencies should not be hidden in prose.',
    ],
  },
};

export function resolveAiDocumentStrategy(
  templateKey?: string | null,
  overrides?: Partial<
    Pick<
      AiDocumentStrategy,
      | 'intentRoute'
      | 'scope'
      | 'sourcePolicy'
      | 'lengthPolicy'
      | 'prioritizeUserInstructions'
    >
  >,
): AiDocumentStrategy {
  const key = templateKey || 'general';
  const override = TEMPLATE_STRATEGIES[key] || {};

  return {
    ...DEFAULT_STRATEGY,
    ...override,
    ...overrides,
    templateKey: key,
    objectives: override.objectives || DEFAULT_STRATEGY.objectives,
    requiredArtifacts:
      override.requiredArtifacts || DEFAULT_STRATEGY.requiredArtifacts,
    optionalArtifacts:
      override.optionalArtifacts || DEFAULT_STRATEGY.optionalArtifacts,
    requiredSections:
      override.requiredSections || DEFAULT_STRATEGY.requiredSections,
    reviewChecks: override.reviewChecks || DEFAULT_STRATEGY.reviewChecks,
    editorSyntaxHints:
      override.editorSyntaxHints || DEFAULT_STRATEGY.editorSyntaxHints,
    intentRoute: overrides?.intentRoute || DEFAULT_STRATEGY.intentRoute,
    scope: overrides?.scope || DEFAULT_STRATEGY.scope,
    sourcePolicy: overrides?.sourcePolicy || DEFAULT_STRATEGY.sourcePolicy,
    lengthPolicy: overrides?.lengthPolicy || DEFAULT_STRATEGY.lengthPolicy,
    prioritizeUserInstructions:
      overrides?.prioritizeUserInstructions ??
      DEFAULT_STRATEGY.prioritizeUserInstructions,
  };
}

export function formatDocumentStrategyForPrompt(
  strategy: AiDocumentStrategy,
): string {
  const lines = [
    'Document strategy:',
    `- Document type: ${strategy.docType}`,
    `- Audience: ${strategy.audience}`,
  ];

  if (strategy.objectives.length > 0) {
    lines.push('- Objectives:');
    for (const item of strategy.objectives) {
      lines.push(`  - ${item}`);
    }
  }

  if (strategy.requiredSections.length > 0) {
    lines.push('- Required sections:');
    for (const item of strategy.requiredSections) {
      lines.push(`  - ${item}`);
    }
  }

  if (strategy.requiredArtifacts.length > 0) {
    lines.push(`- Required artifacts: ${strategy.requiredArtifacts.join(', ')}`);
  }

  if (strategy.optionalArtifacts.length > 0) {
    lines.push(`- Optional artifacts: ${strategy.optionalArtifacts.join(', ')}`);
  }

  if (strategy.reviewChecks.length > 0) {
    lines.push('- Review checks:');
    for (const item of strategy.reviewChecks) {
      lines.push(`  - ${item}`);
    }
  }

  lines.push(`- Intent route: ${strategy.intentRoute}`);
  lines.push(`- Scope: ${strategy.scope}`);
  lines.push(`- Source policy: ${strategy.sourcePolicy}`);
  lines.push(`- Length policy: ${strategy.lengthPolicy}`);
  lines.push(
    `- User instructions have highest priority: ${strategy.prioritizeUserInstructions ? 'yes' : 'no'}`,
  );

  lines.push('- Editor syntax hints:');
  for (const item of strategy.editorSyntaxHints) {
    lines.push(`  - ${item}`);
  }

  return lines.join('\n');
}
